### PR TITLE
add azure_rm_appserviceplan module

### PR DESCRIPTION
**Pull request by yungezz**
##### SUMMARY

add azure_rm_appserviceplan module for azure app service plan.

##### ISSUE TYPE

 - New Module Pull Request

##### ANSIBLE VERSION

```
ansible 2.6
```

##### ADDITIONAL INFORMATION

@zikalino  @yuwzho 

```

```

### CODE REVIEW

**Comment by yungezz** [comment]
ready_for_review

**Comment by yungezz** [comment]
@nitzmahone  @jborean93 @zikalino

**Comment by yungezz** [comment]
@nitzmahone  @jborean93

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+    sku:
+        description:
+            - The pricing tiers, e.g., F1, D1, B1, B2, B3, S1, P1, P1V2 etc.
+            - Please see https://azure.microsoft.com/en-us/pricing/details/app-service/plans/ for more detail.
+            - For linux app service plan, please see https://azure.microsoft.com/en-us/pricing/details/app-service/linux/ for more detail.
+
+    is_linux:
+        description:
+            - Descirbe whether to host webapp on Linux worker.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: Describe*

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+        required: True
+
+    location:
+        description:
+            - Resource location. If not set, location from the resource group will be used as default.
+
+    sku:
+        description:
+            - The pricing tiers, e.g., F1, D1, B1, B2, B3, S1, P1, P1V2 etc.
+            - Please see https://azure.microsoft.com/en-us/pricing/details/app-service/plans/ for more detail.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: Please enclose URLs in `U(https:/....)`

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+            - For linux app service plan, please see https://azure.microsoft.com/en-us/pricing/details/app-service/linux/ for more detail.
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: Enclose URL in `U()`

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+
+author:
+    - "Yunge Zhu(@yungezz)"
+
+'''
+
+EXAMPLES = '''
+    - name: Create a windows app service plan
+      azure_rm_appserviceplan:
+        name: "{{ win_plan_name }}1"
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: Avoid using variables in the examples, use some dummy ones that are somewhat related to the task instead.

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        self.name = None
+        self.location = None
+
+        self.sku = None
+        self.is_linux = None
+        self.number_of_workers = 1
+
+        self.results = dict(
+            changed=False,
+            ansible_facts=dict(azure_appserviceplan=None)
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py:196]: Don't return facts in the `ansible_facts` key unless they are facts about the host that is running the module. Anyway it seems like the facts should be in it's own module not lumped together with this one.

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+    - name: update sku of existing windows app service plan
+      azure_rm_appserviceplan:
+        name: "{{ win_plan_name }}1"
+        resource_group: "{{ plan_resource_group }}"
+        location: "{{ location }}"
+        sku: S2
+'''
+
+RETURN = '''
+azure_appserviceplan:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py:99]: Should this module be returning all this information? Usually we do that for facts modules instead. If you want to keep some values they should just be on the root of the json not in `azure_appserviceplan` as well as documented properly and not shown in the sample.

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+
+                # check if sku changed
+                if self.sku and _normalize_sku(self.sku) != old_response['sku']['size']:
+                    to_be_updated = True
+
+                # check if number_of_workers changed
+                if self.number_of_workers and int(self.number_of_workers) != old_response['sku']['capacity']:
+                    to_be_updated = True
+
+                if self.is_linux and self.is_linux != old_response['reserved']:
```
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py:254]: should this be `reserved`?

File: test/integration/targets/azure_rm_appserviceplan/aliases [hunk 8]
```diff
@@ -0,0 +1,3 @@
+cloud/azure
+posix/ci/cloud/group2/azure
```
Comment (jborean93) [inline test/integration/targets/azure_rm_appserviceplan/aliases]: The aliases have just recently changed, have a look at https://github.com/ansible/ansible/blob/devel/test/integration/targets/azure_rm_loadbalancer/aliases to see how to name this now.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py (hunk 2 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py (hunk 3 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py (hunk 1 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py (hunk 4 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py (hunk 6 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py:99]: fixed to return id.

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py (hunk 5 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py:196]: fixed to return id.

File: lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py (hunk 7 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_appserviceplan.py:254]: fixed

File: test/integration/targets/azure_rm_appserviceplan/aliases (hunk 8 above)
Comment (yungezz) [reply test/integration/targets/azure_rm_appserviceplan/aliases]: fixed

**Pull request merged**